// ask-sdk-jsx-for-apl/dist/lib/responses/apl-document/index.js
"use strict";
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AplDocument = void 0;
const React = __importStar(require("react"));
const server_1 = require("react-dom/server");
const ask_sdk_core_1 = require("ask-sdk-core");
const common_1 = require("../apl/common");
class AplDocument {
    constructor(doc) {
        this.doc = doc;
        this.responseBuilder = ask_sdk_core_1.ResponseFactory.init();
    }
    generateStaticMarkup() {
        if (!this.directiveExists()) {
            server_1.renderToStaticMarkup(React.createElement(common_1.ResponseBuilderCtx.Provider, { value: this.responseBuilder }, this.doc));
        }
    }
    directiveExists() {
        if (this.responseBuilder.getResponse()) {
            return (this.responseBuilder.getResponse().directives || []).find((d) => d.type === 'Alexa.Presentation.APL.RenderDocument');
        }
        else {
            throw 'ResponseBuilder response object can not be null or undefined!';
        }
    }
    getDirective() {
        this.generateStaticMarkup();
        if (this.responseBuilder.getResponse()) {
            return (this.responseBuilder.getResponse().directives || []).find((d) => d.type === 'Alexa.Presentation.APL.RenderDocument');
        }
        else {
            throw 'ResponseBuilder response object can not be null or undefined!';
        }
    }
    getDocument() {
        return this.getDirective().document;
    }
    getDatasources() {
        return this.getDirective().datasources;
    }
}
exports.AplDocument = AplDocument;
