// ask-sdk-jsx-for-apl/dist/lib/responses/apl/common/context.js
"use strict";
/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.APLConsumer = exports.APLProvider = exports.ListComponentConsumer = exports.ListComponentProvider = exports.APLComponentConsumer = exports.APLComponentProvider = exports.MainTemplateConsumer = exports.MainTemplateProvider = exports.ResponseBuilderCtx = void 0;
const react_1 = __importDefault(require("react"));
exports.ResponseBuilderCtx = react_1.default.createContext(null);
_a = react_1.default.createContext(null), exports.MainTemplateProvider = _a.Provider, exports.MainTemplateConsumer = _a.Consumer;
_b = react_1.default.createContext(null), exports.APLComponentProvider = _b.Provider, exports.APLComponentConsumer = _b.Consumer;
_c = react_1.default.createContext(null), exports.ListComponentProvider = _c.Provider, exports.ListComponentConsumer = _c.Consumer;
_d = react_1.default.createContext(null), exports.APLProvider = _d.Provider, exports.APLConsumer = _d.Consumer;
