// ask-sdk-controls/dist/test/CustomSerdeTests.spec.js
"use strict";
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const mocha_1 = require("mocha");
const src_1 = require("../src");
const ControlManager_1 = require("../src/controls/ControlManager");
const GeneralControlIntent_1 = require("../src/intents/GeneralControlIntent");
const ControlHandler_1 = require("../src/runtime/ControlHandler");
const SkillInvoker_1 = require("../src/utils/testSupport/SkillInvoker");
src_1.waitForDebugger();
/*
 * Tests a control that has state (Set<string>) that doesn't automatically serialize with JSON.stringify.
 *
 * The test shows the control changing state and being serialized and deserialized.
 * To accomplish this, the state object implements toJSON() and assign()
 *   - the name and signature for toJSON() is defined by JSON.stringify and called automatically by JSON.stringify
 *   - the name and signature for fromJSON() is defined by the Control framework and used as an alternative to Object.assign() when present.

 */
mocha_1.suite("== Custom Serde ==", () => {
    mocha_1.test('explicit confirmation and disaffirm', async () => {
        // Note: this test demonstrates using testTurn to run a multi-turn scenario with assertions between turns.
        const requestHandler = new ControlHandler_1.ControlHandler(new CustomSerdeControlManager());
        const invoker = new SkillInvoker_1.SkillInvoker(requestHandler);
        await src_1.testTurn(invoker, 'U: ', src_1.TestInput.of(GeneralControlIntent_1.GeneralControlIntent.of({})), 'A:');
        chai_1.expect(requestHandler.getSerializableControlStates().customControl[0]).exist.and.equals('x');
        await src_1.testTurn(invoker, 'U: ', src_1.TestInput.of(GeneralControlIntent_1.GeneralControlIntent.of({})), 'A:');
        chai_1.expect(requestHandler.getSerializableControlStates().customControl[0]).exist.and.equals('y');
    });
});
class CustomSerdeControlManager extends ControlManager_1.ControlManager {
    createControlTree(state) {
        return new CustomSerdeControl({
            id: 'customControl'
        });
    }
}
class CustomSerdeControlState {
    constructor(value) {
        this.value = value;
    }
}
class CustomSerdeControl extends src_1.Control {
    constructor(props, state) {
        super(props.id);
        this.state = state !== null && state !== void 0 ? state : new CustomSerdeControlState(new Set());
    }
    canHandle(input) {
        return true;
    }
    async handle(input, resultBuilder) {
        if (this.state.value.size === 0) {
            this.state.value.add('x');
        }
        else if (this.state.value.size === 1 && this.state.value.has('x')) {
            this.state.value.delete('x');
            this.state.value.add('y');
        }
        return;
    }
    canTakeInitiative(input) {
        return false;
    }
    async takeInitiative(input, resultBuilder) {
        throw new Error("Method not implemented.");
    }
    getSerializableState() {
        // render the set as a simple list.
        return [...this.state.value.keys()];
    }
    setSerializableState(obj) {
        if (obj !== undefined) {
            // refreshes the set from the serialized array
            this.state.value = new Set();
            for (const x of obj) {
                this.state.value.add(x);
            }
        }
    }
    renderAct(act, input, builder) {
        throw new Error("Method not implemented.");
    }
}
