// ask-sdk-controls/dist/test/game_strings.js
"use strict";
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameStrings = void 0;
var GameStrings;
(function (GameStrings) {
    let SlotType;
    (function (SlotType) {
        SlotType["DomainValue"] = "DomainValue";
    })(SlotType = GameStrings.SlotType || (GameStrings.SlotType = {}));
    let Target;
    (function (Target) {
        Target["Name"] = "name";
        Target["AvatarName"] = "avatarName";
        Target["PetName"] = "petName";
        Target["Age"] = "age";
        Target["CharClass"] = "class";
        Target["Species"] = "species";
    })(Target = GameStrings.Target || (GameStrings.Target = {}));
    let ID;
    (function (ID) {
        ID["GameContainer"] = "gameContainer";
        ID["PlayerContainer"] = "playerContainer";
        ID["PlayerName"] = "playerName";
        ID["PlayerAge"] = "playerAge";
        ID["PlayerClass"] = "playerClass";
        ID["PetContainer"] = "petContainer";
        ID["PetName"] = "petName";
        ID["PetSpecies"] = "petSpecies";
    })(ID = GameStrings.ID || (GameStrings.ID = {}));
    let Value;
    (function (Value) {
        Value["Elf"] = "elf";
        Value["Dwarf"] = "dwarf";
        Value["Human"] = "human";
        Value["Cat"] = "cat";
        Value["Dog"] = "dog";
    })(Value = GameStrings.Value || (GameStrings.Value = {}));
})(GameStrings = exports.GameStrings || (exports.GameStrings = {}));
