// ask-sdk-controls/dist/src/constants/Strings.js
"use strict";
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Strings = void 0;
/**
 * The strings associated with various built-ins
 */
var Strings;
(function (Strings) {
    /**
     * Built-in feedback slot values
     */
    let Feedback;
    (function (Feedback) {
        Feedback["Affirm"] = "builtin_affirm";
        Feedback["Disaffirm"] = "builtin_disaffirm";
    })(Feedback = Strings.Feedback || (Strings.Feedback = {}));
    /**
     * Built-in action slot values
     */
    let Action;
    (function (Action) {
        Action["Set"] = "builtin_set";
        Action["Change"] = "builtin_change";
        Action["Select"] = "builtin_select";
        Action["None"] = "builtin_none";
    })(Action = Strings.Action || (Strings.Action = {}));
    /**
     * Built-in target slot values
     */
    let Target;
    (function (Target) {
        Target["It"] = "builtin_it";
        Target["Date"] = "builtin_date";
        Target["StartDate"] = "builtin_start_date";
        Target["EndDate"] = "builtin_end_date";
        Target["DateRange"] = "builtin_date_range";
        Target["Choice"] = "builtin_choice";
        // eslint-disable-next-line id-blacklist
        Target["Number"] = "builtin_number";
    })(Target = Strings.Target || (Strings.Target = {}));
    Strings.Conjunction = 'builtin_conjunction';
    Strings.Tail = 'builtin_tail';
    Strings.Head = 'builtin_head';
    Strings.Preposition = 'builtin_preposition';
})(Strings = exports.Strings || (exports.Strings = {}));
