// ask-sdk-controls/dist/src/intents/AmazonBuiltInIntent.js
"use strict";
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonIntent = void 0;
/**
 * Constants for Alexa's built-in intents.
 *
 * (This is not a complete list)
 */
var AmazonIntent;
(function (AmazonIntent) {
    AmazonIntent["CancelIntent"] = "AMAZON.CancelIntent";
    AmazonIntent["FallbackIntent"] = "AMAZON.FallbackIntent";
    AmazonIntent["HelpIntent"] = "AMAZON.HelpIntent";
    AmazonIntent["MoreIntent"] = "AMAZON.MoreIntent";
    AmazonIntent["NavigateHomeIntent"] = "AMAZON.NavigateHomeIntent";
    AmazonIntent["NavigateSettingsIntent"] = "AMAZON.NavigateSettingsIntent";
    AmazonIntent["NextIntent"] = "AMAZON.NextIntent";
    AmazonIntent["NoIntent"] = "AMAZON.NoIntent";
    AmazonIntent["PageUpIntent"] = "AMAZON.PageUpIntent";
    AmazonIntent["PageDownIntent"] = "AMAZON.PageDownIntent";
    AmazonIntent["PreviousIntent"] = "AMAZON.PreviousIntent";
    AmazonIntent["PauseIntent"] = "AMAZON.PauseIntent";
    AmazonIntent["ResumeIntent"] = "AMAZON.ResumeIntent";
    AmazonIntent["ScrollRightIntent"] = "AMAZON.ScrollRightIntent";
    AmazonIntent["ScrollDownIntent"] = "AMAZON.ScrollDownIntent";
    AmazonIntent["ScrollLeftIntent"] = "AMAZON.ScrollLeftIntent";
    AmazonIntent["ScrollUpIntent"] = "AMAZON.ScrollUpIntent";
    AmazonIntent["StopIntent"] = "AMAZON.StopIntent";
    AmazonIntent["YesIntent"] = "AMAZON.YesIntent";
})(AmazonIntent = exports.AmazonIntent || (exports.AmazonIntent = {}));
