// ask-sdk-controls/dist/src/intents/AmazonBuiltInSlotType.js
"use strict";
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonBuiltInSlotType = void 0;
/**
 * Constants for Alexa's built-in slot types.
 *
 * (This is not a complete list.)
 */
var AmazonBuiltInSlotType;
(function (AmazonBuiltInSlotType) {
    AmazonBuiltInSlotType["DATE"] = "AMAZON.DATE";
    AmazonBuiltInSlotType["DURATION"] = "AMAZON.DURATION";
    AmazonBuiltInSlotType["NUMBER"] = "AMAZON.NUMBER";
    AmazonBuiltInSlotType["FOUR_DIGIT_NUMBER"] = "AMAZON.FOUR_DIGIT_NUMBER";
    AmazonBuiltInSlotType["ORDINAL"] = "AMAZON.Ordinal";
    AmazonBuiltInSlotType["PHONENUMBER"] = "AMAZON.PhoneNumber";
    AmazonBuiltInSlotType["TIME"] = "AMAZON.TIME";
    AmazonBuiltInSlotType["SEARCH_QUERY"] = "AMAZON.SearchQuery";
})(AmazonBuiltInSlotType = exports.AmazonBuiltInSlotType || (exports.AmazonBuiltInSlotType = {}));
