// ask-sdk-controls/dist/src/intents/DateControlIntent.js
"use strict";
/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DateControlIntent = exports.unpackDateControlIntent = void 0;
const IntentUtils_1 = require("../utils/IntentUtils");
const ModelTypes_1 = require("../interactionModelGeneration/ModelTypes");
const BaseControlIntent_1 = require("./BaseControlIntent");
const AmazonBuiltInSlotType_1 = require("./AmazonBuiltInSlotType");
function unpackDateControlIntent(intent) {
    if (!intent.name.startsWith('GeneralControlIntent') && !intent.name.startsWith('DateControlIntent')) {
        throw new Error(`Not an compatible intent: ${intent.name}`);
    }
    let feedback;
    let action;
    let target;
    let amazonDate;
    for (const [name, slot] of Object.entries(intent.slots)) {
        const slotObject = IntentUtils_1.getSlotResolutions(slot);
        const slotValue = slotObject ? slotObject.slotValue : undefined;
        switch (name) {
            case 'feedback':
                feedback = slotValue;
                break;
            case 'action':
                action = slotValue;
                break;
            case 'target':
                target = slotValue;
                break;
            case 'AMAZON.DATE':
                amazonDate = slotValue;
                break;
            // default: ignore content-free slots
        }
    }
    if (amazonDate === undefined) {
        throw new Error("Inconsistency. AMAZON.DATE is expected to be defined. Are canHandle/handle functions in sync");
    }
    return {
        feedback,
        action,
        target,
        'AMAZON.DATE': amazonDate,
    };
}
exports.unpackDateControlIntent = unpackDateControlIntent;
class DateControlIntent extends BaseControlIntent_1.BaseControlIntent {
    static of(slots) {
        return IntentUtils_1.IntentBuilder.of(this.prototype.constructor.name, slots);
    }
    generateIntent() {
        return {
            name: this.name,
            slots: this.generateSlots(),
            samples: [],
        };
    }
    generateSlots() {
        const slots = [
            {
                name: 'feedback',
                type: ModelTypes_1.SharedSlotType.FEEDBACK
            },
            {
                name: 'action',
                type: ModelTypes_1.SharedSlotType.ACTION
            },
            {
                name: 'target',
                type: ModelTypes_1.SharedSlotType.TARGET
            },
            {
                name: 'preposition',
                type: ModelTypes_1.SharedSlotType.PREPOSITION
            },
            {
                name: 'AMAZON.DATE',
                type: AmazonBuiltInSlotType_1.AmazonBuiltInSlotType.DATE
            },
            {
                name: 'head',
                type: ModelTypes_1.SharedSlotType.HEAD
            },
            {
                name: 'tail',
                type: ModelTypes_1.SharedSlotType.TAIL
            }
        ];
        return slots;
    }
    generateSlotTypes() {
        return [];
    }
}
exports.DateControlIntent = DateControlIntent;
